endercore/src/main/java/com/enderio/core/common/blockentity/SyncedBlockEntity.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * create the ClientBoundBlockEntityDataPacket for this BlockEntity
+     * @param completeUpdate if this packet should send all information (this is used for players who started tracking this BlockEntity)
+     * @return the UpdatePacket
+     */
+    @Nullable
+    public ClientboundBlockEntityDataPacket createUpdatePacket(boolean completeUpdate, SyncMode mode) {
+        CompoundTag nbt = new CompoundTag();
+        ListTag listNBT = new ListTag();
+        for (int i = 0; i < dataSlots.size(); i++) {
+            EnderDataSlot<?> dataSlot = dataSlots.get(i);
+            if (dataSlot.getSyncMode() != mode)
+                continue;
+
+            Optional<CompoundTag> optionalNBT = completeUpdate ?
+                Optional.of(dataSlot.toFullNBT()) :
+                dataSlot.toOptionalNBT();
+
+            if (optionalNBT.isPresent()) {
+                CompoundTag elementNBT = optionalNBT.get();
+                elementNBT.putInt("dataSlotIndex", i);
+                listNBT.add(elementNBT);
+            }
+        }
+        if (listNBT.isEmpty())
+            return null;
+        nbt.put("data", listNBT);
+        return new ClientboundBlockEntityDataPacket(this.getBlockPos(), -1, nbt);
+    }
+
+    @Override
+    public void onDataPacket(Connection net, ClientboundBlockEntityDataPacket pkt) {
+        CompoundTag nbt = pkt.getTag();
+
+        ListTag listNBT = nbt.getList("data", Constants.NBT.TAG_COMPOUND);
+        for (Tag tag : listNBT) {
+            CompoundTag elementNBT = (CompoundTag) tag;
+            int dataSlotIndex = elementNBT.getInt("dataSlotIndex");
+            dataSlots.get(dataSlotIndex).handleNBT(elementNBT);
+        }
+    }
+
+    public void addDataSlot(EnderDataSlot<?> slot) {
+        dataSlots.add(slot);
+    }
+
+    /**
+     * Sync the BlockEntity to all tracking players.
+     */
+    @CallOnly(LogicalSide.SERVER)
+    private void sync() {
+        ClientboundBlockEntityDataPacket fullUpdate = createUpdatePacket(true, SyncMode.RENDER);
+        ClientboundBlockEntityDataPacket partialUpdate = getUpdatePacket();
+
+        List<UUID> currentlyTracking = new ArrayList<>();
+
+        getTrackingPlayers().forEach(serverPlayer -> {
+            currentlyTracking.add(serverPlayer.getUUID());
+            if (lastSyncedToPlayers.contains(serverPlayer.getUUID())) {
+                sendPacket(serverPlayer, partialUpdate);
+            } else {
+                sendPacket(serverPlayer, fullUpdate);
+            }
+        });
+        lastSyncedToPlayers.clear();
+        lastSyncedToPlayers.addAll(currentlyTracking);
+    }
+
+    public void sendPacket(ServerPlayer player, @Nullable Packet<?> packet) {
+        if (packet != null)
+            player.connection.send(packet);
+    }
+
+    /**
+     * never call this on client
+     * @return all ServerPlayers tracking this BlockEntity
+     */
+    @CallOnly(LogicalSide.SERVER)
+    private Stream<ServerPlayer> getTrackingPlayers() {
```

Comment:
If I understand this correctly, this will only consider players in the same chunk as the BlockEntity. This could mean a player only one block away will not be returned if on a chunk border. This should probably be changed to a configurable distance as render updates might need to be sent to players a considerable distance away.

Reply:
There are 2 important render distances.
The one for BlockEntitiyRenderers (like the Chest) or the whole rendered world. I  would want to send the updated data to all clients tracking the chunk as some people might want to use the dataSlots for baked models. And I'm fairly certain that the Chunkmap returns all players tracking that chunk based on other useages

Reply:
Thanks, that makes sense